src/snowflake/snowpark/_internal/compiler/plan_compiler.py

```diff
@@ -44,17 +50,33 @@ def should_apply_optimizations(self) -> bool:
 
         current_session = self._plan.session
         return (
-            self._plan.source_plan is not None
-        ) and current_session.cte_optimization_enabled
+            (self._plan.source_plan is not None)
+            and current_session.query_compilation_stage_enabled
+            and current_session.cte_optimization_enabled
+        )
 
     def compile(self) -> Dict[PlanQueryType, List[Query]]:
-        final_plan = self._plan
-        if self.should_apply_optimizations():
-            # apply optimizations
-            final_plan = final_plan.replace_repeated_subquery_with_cte()
-            # TODO: add other optimization steps and code generation step
-
-        return {
-            PlanQueryType.QUERIES: final_plan.queries,
-            PlanQueryType.POST_ACTIONS: final_plan.post_actions,
-        }
+        if self.should_start_query_compilation():
+            # preparation for compilation
+            # 1. make a copy of the original plan
+            logical_plans = [copy.deepcopy(self._plan)]
+            # 2. create a code generator with the original plan
+            query_generator = create_query_generator(self._plan)
+
+            # apply each optimizations if needed
+            if self._plan.session.cte_optimization_enabled:
+                common_sub_dataframe_eliminator = RepeatedSubqueryElimination(
+                    logical_plans, query_generator
+                )
+                logical_plans = common_sub_dataframe_eliminator.apply()
```

Comment:
nit: rename the object?
```suggestion
                repeated_subquery_eliminator = RepeatedSubqueryElimination(
                    logical_plans, query_generator
                )
                logical_plans = repeated_subquery_eliminator.apply()
```

Reply:
updated